Code/Source/Lidar/ROS2LidarSensorComponent.cpp

```diff
@@ -0,0 +1,159 @@
+        m_pointCloudPublisher = ros2Node->create_publisher<sensor_msgs::msg::PointCloud2>(fullTopic.data(), publisherConfig.GetQoS());
+
+        if (m_sensorConfiguration.m_visualise)
+        {
+            auto defaultScene = AZ::RPI::Scene::GetSceneForEntityId(GetEntityId());
+            m_drawQueue = AZ::RPI::AuxGeomFeatureProcessorInterface::GetDrawQueueForScene(defaultScene);
+        }
+    }
+
+    void ROS2LidarSensorComponent::Deactivate()
+    {
+        ROS2SensorComponent::Deactivate();
+        m_pointCloudPublisher.reset();
+    }
+
+    void ROS2LidarSensorComponent::FrequencyTick()
+    {
+        float distance = LidarTemplateUtils::GetTemplate(m_lidarModel).m_maxRange;
+        auto entityTransform = GetEntity()->FindComponent<AzFramework::TransformComponent>(); // TODO - go through ROS2Frame
+        const auto directions = LidarTemplateUtils::PopulateRayDirections(m_lidarModel, entityTransform->GetWorldTM().GetEulerRadians());
+        AZ::Vector3 start = entityTransform->GetWorldTM().GetTranslation();
+        start.SetZ(start.GetZ() + 1.0f);
+        m_lastScanResults = m_lidarRaycaster.PerformRaycast(start, directions, distance);
+        if (m_lastScanResults.empty())
+        {
+            AZ_TracePrintf("Lidar Sensor Component", "No results from raycast");
+            return;
+        }
+
+        if (m_sensorConfiguration.m_visualise)
+        {   // Store points for visualisation purposes, before transformations occur
+            m_visualisationPoints = m_lastScanResults;
+        }
+
+        auto ros2Frame = GetEntity()->FindComponent<ROS2FrameComponent>();
+        auto message = sensor_msgs::msg::PointCloud2();
+        message.header.frame_id = ros2Frame->GetFrameID().data();
+        message.header.stamp = ROS2Interface::Get()->GetROSTimestamp();
+        message.height = 1;
+        message.width = m_lastScanResults.size();
+        message.point_step = sizeof(AZ::Vector3); // TODO - Point Fields can be custom
+        message.row_step = message.width * message.point_step;
+
+        // TODO - a list of supported fields should be returned by lidar implementation
+        std::vector<std::string> point_field_names = { "x", "y", "z"};
+        for (int i = 0; i < point_field_names.size(); i++)
+        {   // TODO - placeholder impl
+            sensor_msgs::msg::PointField pf;
+            pf.name = point_field_names[i];
+            pf.offset = i * 4;
+            pf.datatype = sensor_msgs::msg::PointField::FLOAT32;
+            pf.count = 1;
+            message.fields.push_back(pf);
+        }
+
+        message.data.resize(message.row_step);
+
+        auto lidarTM = entityTransform->GetWorldTM();
+        lidarTM.Invert();
+
+        // TODO - improve performance
+        for (auto& point : m_lastScanResults)
+        {
+            point = lidarTM.TransformPoint(point);
+        }
+
+        memcpy(message.data.data(), m_lastScanResults.data(), message.data.size());
```

Comment:
Is this missing a multiplication against `sizeof <SomeType>`?
It looks like the `message.data` type is some kind of vector, but is it a vector of a 1 byte type.

Reply:
It is correct as it is. The type for PointCloud2 message data field is specified as single-byte type (uint8): 
http://docs.ros.org/en/lunar/api/sensor_msgs/html/msg/PointCloud2.html